core/src/main/java/org/apache/spark/shuffle/external/ExternalShufflePartitionReader.java

```diff
@@ -0,0 +1,78 @@
+package org.apache.spark.shuffle.external;
+
+import org.apache.spark.network.client.TransportClient;
+import org.apache.spark.network.client.TransportClientFactory;
+import org.apache.spark.network.shuffle.protocol.OpenShufflePartition;
+import org.apache.spark.shuffle.api.ShufflePartitionReader;
+import org.apache.spark.storage.ShuffleLocation;
+import org.apache.spark.util.ByteBufferInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.compat.java8.OptionConverters;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Optional;
+
+public class ExternalShufflePartitionReader implements ShufflePartitionReader {
+
+    private static final Logger logger =
+        LoggerFactory.getLogger(ExternalShufflePartitionReader.class);
+
+    private final TransportClientFactory clientFactory;
+    private final String appId;
+    private final int shuffleId;
+    private final int mapId;
+
+    public ExternalShufflePartitionReader(
+            TransportClientFactory clientFactory,
+            String appId,
+            int shuffleId,
+            int mapId) {
+        this.clientFactory = clientFactory;
+        this.appId = appId;
+        this.shuffleId = shuffleId;
+        this.mapId = mapId;
+    }
+
+    @Override
+    public InputStream fetchPartition(int reduceId, Optional<ShuffleLocation> shuffleLocation) {
+        assert shuffleLocation.isPresent() && shuffleLocation.get() instanceof ExternalShuffleLocation;
+        ExternalShuffleLocation externalShuffleLocation = (ExternalShuffleLocation) shuffleLocation.get();
+        logger.info(String.format("Found external shuffle location on node: %s:%d",
+                externalShuffleLocation.getShuffleHostname(),
+                externalShuffleLocation.getShufflePort()));
+        String hostname = externalShuffleLocation.getShuffleHostname();
+        int port = externalShuffleLocation.getShufflePort();
+
+        OpenShufflePartition openMessage =
+            new OpenShufflePartition(appId, shuffleId, mapId, reduceId);
+        TransportClient client = null;
+        try {
+            client = clientFactory.createUnmanagedClient(hostname, port);
+            String requestID = String.format(
+                    "read-%s-%d-%d-%d", appId, shuffleId, mapId, reduceId);
+            client.setClientId(requestID);
+            logger.info("clientid: " + client.getClientId() + " " + client.isActive());
+
+            ByteBuffer response = client.sendRpcSync(openMessage.toByteBuffer(), 60000);
```

Comment:
here you're always fetching the remote block into memory, you won't use fetch-to-disk at all.

Also you wont' do any pipelining of requests.  You'll request one block, fetch it all into memory, consume the entire block, and only then start the request for the next block.  I mentioned this on matt's work on the api too, but wanted to mention it again here.

I understand if this doesn't go in right away, but we want to make sure we design things so this is possible, at least.

Reply:
How is this done currently? Does the prior implementation do fetch to disk directly?

Reply:
Its quite a bit more complicated.

(1) `ShuffleBlockFetcherIterator` looks at the block size to determine whether or not to fetch it to disk:
https://github.com/apache/spark/blob/dc2da72100811988ee1b31190f219b620f88f8de/core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala#L253-L261
(btw it also uses that size info to know how many requests to pipeline)

(2) eventually that gets down to `OneForOneBlockFetcher` here:
https://github.com/apache/spark/blob/dc2da72100811988ee1b31190f219b620f88f8de/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java#L123-L127

the msg it sends is slightly different, but more or less its still just saying "I want shuffle block X".

(3) The receiving end still looks at that msg to figure out what shuffle block is requested, and then streams back that data.  The file on disk is still pushed directly to the socket, without copying into memory in the JVM, regardless of whether the receiving end wants them on disk or not.  But the header it sends back is slightly different.

(4) because the response is a `StreamResponse`, the `TransportResponseHandler` puts an interceptor in the frame decoder, so that rather than buffering up the entire msg, it just send the bytes in small chunks on to the callback.

https://github.com/apache/spark/blob/dc2da72100811988ee1b31190f219b620f88f8de/common/network-common/src/main/java/org/apache/spark/network/client/TransportResponseHandler.java#L209-L220

Also see `MessageEncoder` and how it uses `isBodyInFrame`.

Going back up to OneForOneBlockFetcher, that callback is a `DownloadCallback` which just writes bytes to a file:

https://github.com/apache/spark/blob/dc2da72100811988ee1b31190f219b620f88f8de/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java#L155-L172

and when its fetched all the data, it gives a view of the bytes in that file as a ManagedBuffer, so other parts of spark don't care whether its on disk or in memory (the details here get complicated by encryption).

Reply:
Does implementing this require changes at the API layer? Or is it only on the implementation of our API?

Reply:
pipelining would almost certainly require API changes (but probably just pushing down a bit of the logic here: https://github.com/mccheah/spark/pull/4/files#r246172407

Fetch-to-disk I think mostly just requires a change to the implementation.  But it does require knowledge of the size of the shuffle block that is about to be fetched (or at least, whether or not its bigger than the limit) -- I'm not sure I follow which parts of that will be changed in the api yet.

Reply:
btw another thing I was thinking was if you could push the api change down even lower, so that you only change the `ShuffleClient` used, but keep the `ShuffleBlockFetcherIterator`.  You'd have to change the arguments to `ShuffleClient.fetchBlocks` (and I'm not exactly sure how atm) but otherwise maybe you want to reuse the rest of `ShuffleBlockFetcherIterator`?

Reply:
To clarify, is the suggestion to make `ShuffleClient` the API that developers would customize to work with pluggable backends?